src/Microsoft.OData.Core/UriParser/Binders/KeyBinder.cs

```diff
@@ -112,11 +197,12 @@ private KeyPropertyValue BindKeyPropertyValue(NamedValue namedValue, IEdmEntityT
             }
             else
             {
-                keyProperty = collectionItemEntityType.Key().Where(k => string.CompareOrdinal(k.Name, namedValue.Name) == 0).SingleOrDefault();
+                keyProperty = keys.Where(k => string.CompareOrdinal(k.Key, namedValue.Name) == 0).SingleOrDefault().Value;
 
                 if (keyProperty == null)
                 {
-                    throw new ODataException(ODataErrorStrings.MetadataBinder_PropertyNotDeclaredOrNotKeyInKeyValue(namedValue.Name, collectionItemEntityType.ODataFullName()));

```

Comment:
there is an exception thrown from BindKeyValues to take care of scenario where segment cannot be bound to declared keys.